cpp/arcticdb/version/symbol_list.cpp

```diff
@@ -6,292 +6,738 @@
+            ARCTICDB_DEBUG(log::symbol(), "Reading numeric symbol {}", num_id);
+            return safe_convert_to_numeric_id(num_id);
+    } else {
+        auto sym = std::string(seg.string_at(row_id, column).value());
+        ARCTICDB_DEBUG(log::symbol(), "Reading string symbol '{}'", sym);
+        return {std::move(sym)};
+    }
 }
 
-    void SymbolList::write_journal(const std::shared_ptr<Store>& store, const StreamId& symbol, std::string action) {
-        SegmentInMemory seg{journal_stream_descriptor(action, symbol)};
-        util::variant_match(symbol,
-            [&seg] (const StringId& id) {
-                seg.set_string(0, id);
-            },
-            [&seg] (const NumericId& id) {
-                seg.set_scalar<uint64_t>(0, id);
-            });
+DataType get_symbol_data_type(const SegmentInMemory& seg) {
+    const auto& field_desc = seg.descriptor().field(0);
+    auto data_type =  field_desc.type().data_type();
 
-        seg.end_row();
-        try {
-            store->write_sync(KeyType::SYMBOL_LIST, 0, StreamId{ action }, IndexValue{ symbol }, IndexValue{ symbol },
-                std::move(seg));
-        } catch ([[maybe_unused]] const arcticdb::storage::DuplicateKeyException& e)  {
-            // Both version and content hash are fixed, so collision is possible
-            ARCTICDB_DEBUG(log::storage(), "Symbol list DuplicateKeyException: {}", e.what());
-        }
+    missing_data::check<ErrorCode::E_UNREADABLE_SYMBOL_LIST>(data_type== DataType::UINT64 || data_type == DataType::ASCII_DYNAMIC64,
+        "The symbol list contains unsupported symbol type: {}", data_type);
+
+    return data_type;
+}
+
+std::vector<SymbolListEntry> read_old_style_list_from_storage(const SegmentInMemory& seg) {
+    std::vector<SymbolListEntry> output;
+    if(seg.empty())
+        return output;
+
+    const auto data_type = get_symbol_data_type(seg);
+
+    for(auto row : seg)
+        output.emplace_back(stream_id_from_segment(data_type, seg, row.row_id_, 0), 0, 0, ActionType::ADD);
+
+    return output;
+}
+
+std::vector<SymbolListEntry> read_new_style_list_from_storage(const SegmentInMemory& seg) {
+    std::vector<SymbolListEntry> output;
+    if(seg.row_count() == 0)
+        return output;
+
+    const auto data_type = get_symbol_data_type(seg);
+
+    // Because we need to be backwards compatible with the old style symbol list, the additions and deletions are
+    // in separate columns. The first three columns are the symbol, version and timestamp for the additions, and the
+    // next three are the same for the deletions. Old-style symbol lists will ignore everything but the first column
+    // which will mean that they can't do any conflict resolution but will get the correct data.
+    for(auto i = 0L; i < seg.column(0).row_count(); ++i) {
+        util::check(seg.column(0).row_count() == seg.column(1).row_count() && seg.column(0).row_count() == seg.column(2).row_count(),
```

Comment:
Having an enum with the column names would be more readable than using "magic numbers"